Turn main dialog into regular dialog

er-vin opened the pull request:
After discussion with @jancborchardt and taking into account feedback we got during the 3.0 cycle, we decided to turn the tray dialog into a regular dialog (so resizable and with a window title). There's been needs expressed for the ability to move it around, resize it and not have it disappear when it looses focus for some tasks.

This mainly revolves around needs to compare information with another window (say the file manager) or when needing to type some of the information displayed somewhere else.

It also allows to reduce a lot the amount of platform specific code in the systray. The geometry of the dialog is saved and restored in a similar fashion to the settings dialog.

Fix #2646 
Fix #2607
Fix #2569
Fix #2425 (note it's very likely the same for windows when that option is enabled there, already got it when on windows the systray icon is in the pop over instead of straight in the bar)
Should fix #2643
Should fix #2332

There are others where it was mostly in the comments so harder to find. There are also others we closed in the past because "it is a design decision" but I don't think that's sustainable anymore.

Fix in part #2638 (also need extra fixes for dpi issues which still remain).
Talking about hidpi, that doesn't fix those but it's a step in the right direction to be able to address them fully in another PR: #2295 or #1079 (last one is a bit confused and actually about several platforms, there are a couple more in the list)

I also suspect it might make some of the crashes with seen on Windows to go away (although I don't have the right combinations of hardware and drivers to test) since the path between the application and the window manager going through the GPU changes (it's just not the same for a non rectangular surface with no window decoration and a dialog).

As for looks, it's not a huge change in the end, it looses rounded corners and gains a titlebar, most of the identity is retained, example on Windows follows.

Before (Windows) | After (Windows):
-|-
![Screenshot_20201119_100159](https://user-images.githubusercontent.com/1411800/99644483-49dd7d00-2a4e-11eb-8f8a-e197ee5109ec.png) | ![Screenshot_20201119_095242](https://user-images.githubusercontent.com/1411800/99643634-37167880-2a4d-11eb-851e-5288965c53a3.png)

Before (Mac) | After (Mac)
-|-
![Screenshot_20201119_142024](https://user-images.githubusercontent.com/1411800/99671476-6095cb00-2a72-11eb-9d3d-08c658b5c1f3.png) | ![Screenshot_20201119_141621](https://user-images.githubusercontent.com/1411800/99671082-d2b9e000-2a71-11eb-90b4-c2ae0e6b4596.png)

er-vin commented:
/rebase

splitt3r commented:
Does it still look the same after converting it? The OneDrive windows e.g. has a fixed size and position too. Also close on focus is pretty common isn't it?

er-vin commented:
> Does it still look the same after converting it?

No it doesn't. It now behaves more like a regular dialog. Corners and title bar are in control of the platform.

> The OneDrive windows e.g. has a fixed size and position too.

Which is something you can afford when you're deeply integrated with the platform. Less so when you're not and you got to maintain that on three++ platforms with a single tiny team.

> Also close on focus is pretty common isn't it?

To me it's not, but we got quite a few complaints at it revolving around the activity stream. This one being transient is not necessarily practical for all the use cases people come up with. I admit I was surprised. :-)

splitt3r commented:
> > Does it still look the same after converting it?
> 
> No it doesn't. It now behaves more like a regular dialog. Corners and title bar are in control of the platform.
> 
> > The OneDrive windows e.g. has a fixed size and position too.
> 
> Which is something you can afford when you're deeply integrated with the platform. Less so when you're not and you got to maintain that on three++ platforms with a single tiny team.
> 
> > Also close on focus is pretty common isn't it?
> 
> To me it's not, but we got quite a few complaints at it revolving around the activity stream. This one being transient is not necessarily practical for all the use cases people come up with. I admit I was surprised. :-)

Of course i get your points and i see the hassle of maintaining the code for different OSes. But i also think the users which complain about things on some Linux distros e.g. on GitHub often have some special use cases. I think normal end users would expect the Nextcloud app to look as nice as OneDrive or Dropbox and behave as easy as them. But in the end you as the Nextcloud company know best. Keep up the great work. I really like the tray window from @DominiqueFuchs.

er-vin commented:
> Of course i get your points and i see the hassle of maintaining the code for different OSes. But i also think the users which complain about things on some Linux distros e.g. on GitHub often have some special use cases. I think normal end users would expect the Nextcloud app to look as nice as OneDrive or Dropbox and behave as easy as them.

Well, we got feedback from Mac people as well. So yeah...

> But in the end you as the Nextcloud company know best. Keep up the great work. I really like the tray window from @DominiqueFuchs.

Well, the big bulk of it is still there of course. It's not like we removed the content of it, it's mostly how it ties to the platform that we changed.

jancborchardt commented:
This is of course a compromise – ideally it still would look great while working great as well. As @er-vin describes there’s drawbacks.

@er-vin could you post a screenshot of before & after, and cc the people from the issues which would be fixed with this to see if this move actually makes sense to fix it?
Cause as we talked about as well – one of my worries is as well that if we cater too much to the crowd which knows how to use Github, we actually make it look strange for people who are used to simple solutions. 

Also just for reference @karlitschek because this will change the design in a way which could be seen as a step back.

allexzander commented on src/gui/systray.cpp:
```diff
@@ -190,19 +191,34 @@ void Systray::setPauseOnAllFoldersHelper(bool pause)
 }
 
-Q_INVOKABLE void Systray::setOpened()
+void Systray::setWindowVisible(bool value)
```
A question, regarding naming conventions. Shouldn't we use `void Systray::set**Is**WindowVisible(bool value)` since we have the variable named _**is**WindowVisible not just _windowVisible?

er-vin replied:
This is following Qt API convention there, so windowVisible / isWindowVisible / setWindowVisible for a bool property. As for the member variable I tend to keep the is as well, there's no clear convention on that though.

allexzander commented on src/libsync/configfile.cpp:
```diff
@@ -262,6 +344,24 @@ void ConfigFile::restoreGeometry(QWidget *w)
+void ConfigFile::saveGeometry(QWindow *w)
+{
+#ifndef TOKEN_AUTH_ONLY
+    ASSERT(!w->objectName().isNull());
```
@er-vin Isn't it necessary to check **w** is not null prior to calling **objectName()** on it? Also, maybe it makes sense to use **Q_ASSERT** instead of **ASSERT**?

er-vin replied:
Admittedly I went along with what was in the saveGeometry(QWidget *w) overload. I'd rather keep improvements like the one you propose here for a different PR or such (where both would be changed together).

allexzander commented on src/libsync/configfile.cpp:
```diff
@@ -143,6 +145,86 @@ bool copy_dir_recursive(QString from_dir, QString to_dir)
+    const auto frameHeight = 20;
+
+    if (!restoredGeometry->intersects(availableGeometry)) {
+        restoredGeometry->moveBottom(qMin(restoredGeometry->bottom(), availableGeometry.bottom()));
```
@er-vin Can this be achieved with just calling **translate** or **moveToo** method of **QRect**? Also, maybe it's better to just put the window into the center of the screen in case we get into this condition?

er-vin replied:
Yes, but this would actually lead to more verbose code. I used exactly the code used in Qt for the widget variant though.

allexzander commented on src/libsync/configfile.cpp:
```diff
@@ -143,6 +145,86 @@ bool copy_dir_recursive(QString from_dir, QString to_dir)
+    QDataStream stream(geometry);
+    stream.setVersion(QDataStream::Qt_5_12);
+
+    const quint32 magicNumber = 0x430A7465;
```
@er-vin Move this into static const since it is hardcoded in two methods?

er-vin replied:
Anonymous namespace and constexpr but yes, good point. I'll add a commit for that.

er-vin replied:
Done

er-vin commented:
> This is of course a compromise – ideally it still would look great while working great as well. As @er-vin describes there’s drawbacks.

Yes, clearly it's compromising here.
 
> @er-vin could you post a screenshot of before & after, and cc the people from the issues which would be fixed with this to see if this move actually makes sense to fix it?

I updated the description of the PR to reference issues it fixes or will help with down the line.

> Cause as we talked about as well – one of my worries is as well that if we cater too much to the crowd which knows how to use Github, we actually make it look strange for people who are used to simple solutions.

Sure, let me remind one thing though, I waited several months after we rolled out the new GUI officially before doing any move there. I also pushed back the first issues opened during that period mostly out of "look, it's the new design, deal with it". After that phase we're still having new issues being open or comments in other issues which are quite a bit more reasonable, that's why I believe we might have to do that move now.

And finally, a couple of things I didn't necessarily discuss with @jancborchardt but which came back to me while updating the description above:
 * we have HiDPI issues on Windows (among others) and some of the code changes in this PR will help tackle them (because it fixes layout issues that you see only on HiDPI or when the window is resizable);
 * the 3.0 client is very sensitive to driver issues on Windows, and I suspect that it might help to have a regular Window in our situation (of course, let me stress it **I'm not sure** it does in the end, I didn't get those issues and I lack the hardware to test).

karlitschek commented:
I think it would be good to clearly check how this affects the look and feel on all platforms. Priorities ordered by the number of affected users. So priority 1 is Windows, priority 2 is Mac, priority 3 is Linux with KDE/GNOME. Low priority is good behaviours on non standard setups like 'focus follows mouse'.  What is a good way to evaluate this? @jancborchardt @er-vin

er-vin commented:
> I think it would be good to clearly check how this affects the look and feel on all platforms. Priorities ordered by the number of affected users. So priority 1 is Windows, priority 2 is Mac,

I added the Windows and Mac screenshots for before / after.

But I agree they have higher priority, that's why we didn't budge on feedback until that were problems closer to number 1 and 2 platforms.

> priority 3 is Linux with KDE/GNOME.

In term of look that's fairly close to what you get on Mac (different buttons but otherwise similar coloring and size by default for the title bars). In case you wonder why I didn't produce the screenshots for those.

> Low priority is good behaviours on non standard setups like 'focus follows mouse'. What is a good way to evaluate this? @jancborchardt @er-vin

Not sure what you meant with that question. As far as focus follows mouse is concerned, currently it's just outright unusable for those and we have no way of knowing if a user enabled it or not.

jancborchardt commented:
@karlitschek could you check the before & after screenshots in @er-vin top post if the tradeoff is acceptable from your view? As said I would of course prefer it without heading, but @er-vin makes the calls on cross-platform issues and workload.

karlitschek commented:
To be honest I'm not sure. This 'fixes' only some no common and not standard scenarios. But makes everything significantly uglier and weird.

DominiqueFuchs commented:
Just a facile opinion, as my review was requested a few times now: I get the intention of this, but personally would not go for it. May be biased bc I came up with this and heavily advocate the current design/behavior to satisfy the (as of today) common sense of ux intuition for _normal_ users.

- Would it be more work to make this really nice/usable for everyone? Yes, obviously.
- Do I think one should avoid that work (resulting in: longer term fixing up bc the development resources on this team are minimal) and instead just revert it to a plain old windowed experience? No, even though this may block some more special users from this new major version for some time.

Just my 2 cents.

er-vin commented:
Okidoki, let's go for status quo then.

tanghus commented:
I'm unsure how this issue ended? Will it stay as the non-standard window, or will it be changed/reverted? It does look very weird on Plasma with a dark theme. I'm currently on 3.0.3 on Kubuntu 20.10.
But I may just be one of the "not normal" users ;)
![Screenshot_20201130_094236](https://user-images.githubusercontent.com/533222/100587066-643a1500-32f0-11eb-8451-cc6dd425c16a.png)
.

jancborchardt commented:
@tanghus we are going to stick with the custom dialog since the title bar looks very off compared to well-known solutions like Dropbox, Google Drive and the like.

There are a lot of things to be improved with the dialog – like the issues referenced above – and we will continue working on improving and fixing things. :)

tanghus commented:
> compared to well-known solutions like Dropbox, Google Drive and the like.

:thinking: ... argument very reluctantly accepted :wink:

juampii commented:
Any chance the new dialog respect themes? qt. For example a dark theme in plasma 5 (or even windows) looks really ugly compared to previous versions. Also its harder to "see" whats is going on when some conflict emerges. If not can you please link anything to read about this change and why? Regards

janvlug commented:
> we are going to stick with the custom dialog since the title bar looks very off compared to well-known solutions like Dropbox, Google Drive and the like.

I'm a standard Fedora 33 GNOME on Wayland user. I have to say that since the GUI update accessing the Nextcloud client is barely working for me. Also the design of the main window looks very out of place in the general GNOME look and feel. Would it be possible to turn the main dialog into a regular window for the Linux platform? Some issues that I experience due to the fact of the non-standard GUI:
* [No tray icon shown](https://github.com/nextcloud/desktop/issues/2783)
* Unable to resize the main window (now difficult to read easily the messages)
* Unable to move the main window, it is just dropped on a unwanted place on my desktop.
* When I select "Open main dialog" from the tray icon, the main window sometimes shows briefly and disappears again.

Please, note that I appreciate all the work of the Nextcloud team and the above is intended as constructive feedback. With the current behaviour of the Nextcloud client main window on Fedora, a regular Fedora user is not able to use Nextcloud client, which is unfortunate. I guess several of these issues apply for Ubuntu users as well.

I doubted whether I should comment on this closed issue or that I should have created a new issue with the request to have a regular dialog as main window on Linux. Please let me know if you would prefer that, and I will create a new issue.
